File: 032_longest_valid_parentheses.java
public class Solution {
    public int longestValidParentheses(String s) {
        int maxans = 0;
        Stack<Integer> stack = new Stack<>();
        stack.push(-1);
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                stack.pop();
                if (stack.size() == 0) {
                    stack.push(i);
                } else {
                    // calculate the length of the currently encountered valid string of parentheses
                    maxans = Math.max(maxans, i - stack.peek());
                }
            }
        }
        return maxans;
    }
}

File: python/032_longest_valid_parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        stack = [-1]
        ans = 0
        for i in range(len(s)):
            if s[i] == '(':
                stack.append(i)
            else:
                if stack:
                    stack.pop()
                if not stack:
                    # since stack is empty, indicateing previous parentheses
                    # are valid, and it shows s[:i] are valid
                    # so we can push current index to stack
                    stack.append(i)
                else:
                    ans = max(ans, i - stack[-1])
        return ans
